### PR TITLE
fixes #19782; make `?` concatenate query data

### LINKED ISSUE #19782: uri.`?` doesn't concatenate
I expected `?` to not overwrite the query, but to merge. The docs of the proc use the word "concatenate" which implies that this is how the proc should work, but the implementation overwrites the `query` instead of concatenating.

### Example
```nim
import uri
let foo = parseUri("https://google.com/?page=10")
echo(foo ? {"test": "42"})
```

### Current Output

```
https://google.com/?test=42
```

### Expected Output
```
https://google.com/?page=10&test=42
```

### Possible Solution

Workaround is:

```nim
echo(foo ? (toSeq(foo.query.decodeQuery()) & @({"test": "42"})))
```

We can either change the semantics behind a feature flag (to avoid breaking existing code) or update the docs.

### PR DESCRIPTION
**Pull request by Nmerryman**
fixes #19782

The bug involved caused the uri.`?` concatenate function to overwrite current query data. The new implementation adds a check too ensure the data is not lost.
Also added some tests to account for these situations

Something to consider is that these tests expect the query order to stay the same.

### CODE REVIEW

File: lib/pure/uri.nim [hunk 1]
```diff
@@ -483,7 +483,11 @@ func `?`*(u: Uri, query: openArray[(string, string)]): Uri =
     let foo = parseUri("https://example.com") / "foo" ? {"bar": "qux"}
     assert $foo == "https://example.com/foo?bar=qux"
   result = u
-  result.query = encodeQuery(query)
+  # Makes sure to keep previous query if found
+  if result.query != "":
+    result.query = result.query & "&" & encodeQuery(query)
```
Comment (ghost) [inline lib/pure/uri.nim]: Why make a new string instead of just `add`ing the & and encodeQuery(query) ?

Comment (ghost) [reply lib/pure/uri.nim]:
```suggestion
    result.query.add '&'
    result.query.add encodeQuery(query)
```

Something like this (of course you can do this in 1 line as well)

Comment (Nmerryman) [reply lib/pure/uri.nim]: Nothing wrong with using add instead, I just sort of forgot about it as I wanted to make it work first before worrying about optimizations. I guess python habits got in the way.

Comment (Nmerryman) [reply lib/pure/uri.nim]: I was curious and ran some benchmarks real quick and my results suggest that the original method via setting is more than 2x times faster than using `and`. [Playground benchmark](https://play.nim-lang.org/#ix=3Y3n) 
Unless there is a different reason to you what you suggested, I think I will revert it to how it was originally.

Comment (ghost) [reply lib/pure/uri.nim]: @Nmerryman I also ran some benchmarks with benchy, see this:
```nim
import std/uri

# presetting
# for a in 0 ..< runs:
#   adding[a].query.add("a")
#   setting[a].query = "a"

import benchy

timeIt "adding":
  var q = initUri()
  for i in 0..1000:
    q.query.add("aaaaaaaaaaa")
    q.query.add("aaaaaaaaaaa")
    keep q

timeIt "setting":
  var q = initUri()
  for i in 0..1000:
    q.query = q.query & "aaaaaaaaaaa" & "aaaaaaaaaaa"
    keep q
```

here your method is much slower. The main thing that code does is add stuff to _existing_ query, so this benchmark shows it better than your one. And apart from the time there's also amount of allocations done.

Comment (ghost) [reply lib/pure/uri.nim]: Anyway, my comment was merely a suggestion, if you consider your way to be better, go ahead with it.

Comment (ghost) [reply lib/pure/uri.nim]: Also you can check the memory allocation usage:
```nim
import std/uri

when true:
  var q = initUri()
  for i in 0..1000:
    q.query.add("aaaaaaaaaaa")
    q.query.add("aaaaaaaaaaa")
  doAssert q.query.len > 5000

when false:
  var q = initUri()
  for i in 0..1000:
    q.query = q.query & "aaaaaaaaaaa" & "aaaaaaaaaaa"
  doAssert q.query.len > 5000
```
  
When the first branch is true we get this (with `-d:danger --gc:arc -d:useMalloc`):

`total heap usage: 12 allocs, 12 frees, 45,153 bytes allocated`

In second case (if we make the first branch `when false` and the second one `when true`) we get:

`total heap usage: 1,001 allocs, 1,001 frees, 11,042,031 bytes allocated`

**Review by dom96 — APPROVED** [review]
LGTM but isn't this a breaking change?

**Comment by ringabout** [comment]
ref https://github.com/nim-lang/Nim/issues/19782#issuecomment-1498906611

**Comment by Araq** [comment]
Needs at least a changelog entry.

**Comment by ringabout** [comment]
ref https://github.com/nim-lang/Nim/pull/22784#issuecomment-1744890414

**Pull request closed without merging**